x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/DecryptionPacketsInputStream.java

```diff
@@ -0,0 +1,158 @@
+ * that have been used during encryption must also be used for decryption, otherwise
+ * decryption will fail.
+ * <p>
+ * This implementation buffers the encrypted packet in memory. The maximum packet size it can
+ * accommodate is {@link EncryptedRepository#MAX_PACKET_LENGTH_IN_BYTES}.
+ * <p>
+ * This implementation does not support {@code mark} and {@code reset}.
+ * <p>
+ * The {@code close} call will close the decryption input stream and any subsequent {@code read},
+ * {@code skip}, {@code available} and {@code reset} calls will throw {@code IOException}s.
+ * <p>
+ * This is NOT thread-safe, multiple threads sharing a single instance must synchronize access.
+ *
+ * @see EncryptionPacketsInputStream
+ */
+public final class DecryptionPacketsInputStream extends ChainingInputStream {
+
+    private final InputStream source;
+    private final SecretKey secretKey;
+    private final int nonce;
+    private final int packetLength;
+    private final byte[] packet;
+    private final byte[] iv;
+
+    private boolean hasNext;
+    private long counter;
+
+    /**
+     * Computes and returns the length of the plaintext given the {@code ciphertextLength} and the {@code packetLength}
+     * used during encryption.
+     * Each ciphertext packet is prepended by the Initilization Vector and appended the Authentication Tag.
+     * Decryption is 1:1, and the ciphertext is not padded, but stripping away the IV and the AT amounts to a shorter
+     * plaintext compared to the ciphertext.
+     *
+     * @see EncryptionPacketsInputStream#getEncryptionLength(long, int)
+     */
+    public static long getDecryptionLength(long ciphertextLength, int packetLength) {
+        long encryptedPacketLength =
+                packetLength + EncryptedRepository.GCM_TAG_LENGTH_IN_BYTES + EncryptedRepository.GCM_IV_LENGTH_IN_BYTES;
+        long completePackets = ciphertextLength / encryptedPacketLength;
+        long decryptedSize = completePackets * packetLength;
+        if (ciphertextLength % encryptedPacketLength != 0) {
+            decryptedSize += (ciphertextLength % encryptedPacketLength) - EncryptedRepository.GCM_IV_LENGTH_IN_BYTES
+                    - EncryptedRepository.GCM_TAG_LENGTH_IN_BYTES;
+        }
+        return decryptedSize;
+    }
+
+    public DecryptionPacketsInputStream(InputStream source, SecretKey secretKey, int nonce, int packetLength) {
+        this.source = Objects.requireNonNull(source);
+        this.secretKey = Objects.requireNonNull(secretKey);
+        this.nonce = nonce;
+        if (packetLength <= 0 || packetLength >= EncryptedRepository.MAX_PACKET_LENGTH_IN_BYTES) {
+            throw new IllegalArgumentException("Invalid packet length [" + packetLength + "]");
+        }
+        this.packetLength = packetLength;
+        this.packet = new byte[packetLength + EncryptedRepository.GCM_TAG_LENGTH_IN_BYTES];
+        this.iv = new byte[EncryptedRepository.GCM_IV_LENGTH_IN_BYTES];
```

Comment:
It looks this is a field just so we can avoid allocating a 12 byte array as a local var in `decrypt`.
That seems like the wrong tradeoff to me.

Reply:
I share your apprehension of instance variables to avoid allocations.

I have pushed 8a0773aba5f8a9436e993a8 which reuses the `packet` byte array, i.e. it first reads the IV into the packet buffer to create the `packetCipher` and then it reads the rest of the packet and decrypts it in-place.

If you deem this sketchy, I will allocate the IV byte[] inside the `decrypt` method.

Reply:
I would like to avoid allocating the packet array large enough to contain both the IV and the rest of the packet because I worry there might be performance issues when the actual decryption does not start from an aligned address (the +12 offset).